src/libstd/sync/condvar.rs

```diff
@@ -167,6 +163,37 @@ impl Condvar {
         }
     }
 
+    /// Wait on this condition variable for a notification, timing out after a
+    /// specified duration.
+    ///
+    /// The semantics of this function are equivalent to `wait_timeout` except
+    /// that the implementation will repeatedly wait while the duration has not
+    /// passed and the provided function returns `false`.
+    // FIXME this doesn't handle the poison case at all if the user wanted to
+    // unwrap the Err and keep on waiting
+    #[unstable]
+    pub fn wait_timeout_with<'a, T, F>(&self,
```

Comment:
This is an interesting precedent (`_with` variants) as it means we one day may have 6 methods for waiting: `{wait,wait_for,wait_until}{,_with}`. Given that it does seem like there may be an opportunity to unify these methods. I don't have many ideas off the top of my head, however, and this is `#[unstable]`, so it's largely just food-for-thought.